### ISSUE TITLE
BUG_ON in __bch2_btree_node_write during fsck [46b2c55]

### ISSUE THREAD

**Issue by RAOF** [koverstreet/bcachefs#421]
This is also hit by the kernel fsck.

<pre><font color="#859900"><b>ubuntu@ubuntu</b></font>:<font color="#268BD2"><b>~/bcachefs-tools</b></font>$ ./bcachefs version
bcachefs tool version v0.1-530-g46b2c55
</pre>

<pre><font color="#859900"><b>ubuntu@ubuntu</b></font>:<font color="#268BD2"><b>~/bcachefs-tools</b></font>$ sudo ./bcachefs show-super /dev/nvme0n1 
External UUID:                  a4b165a2-556b-4650-88b8-d90f8ee4b473
Internal UUID:                  7eefb620-ddab-4859-9449-9ca659257331
Device index:                   5
Label:                          
Version:                        new_data_types
Oldest version on disk:         subvol_dirent
Created:                        Tue Nov  2 06:56:16 2021
Sequence number:                243
Superblock size:                37224
Clean:                          0
Devices:                        6
Sections:                       journal,members,crypt,replicas_v0,disk_groups,clean,journal_seq_blacklist,counters
Features:                       lz4,journal_seq_blacklist_v3,reflink,new_siphash,inline_data,new_extent_overwrite,incompressible,btree_ptr_v2,extents_above_btree_updates,btree_updates_journalled,reflink_inline_data,new_varint,journal_no_flush,alloc_v2,extents_across_btree_nodes
Compat features:                alloc_info,alloc_metadata,extents_above_btree_updates_done,bformat_overflow_done

Options:
  block_size:                   4k
  btree_node_size:              256k
  errors:                       continue [ro] panic 
  metadata_replicas:            2
  data_replicas:                2
  metadata_replicas_required:   1
  data_replicas_required:       1
  encoded_extent_max:           64k
  metadata_checksum:            none [crc32c] crc64 xxhash 
  data_checksum:                none [crc32c] crc64 xxhash 
  compression:                  none [lz4] gzip zstd 
  background_compression:       [none] lz4 gzip zstd 
  str_hash:                     crc32c crc64 [siphash] 
  metadata_target:              ssd
  foreground_target:            ssd
  background_target:            rotational
  promote_target:               ssd
  erasure_code:                 0
  inodes_32bit:                 1
  shard_inode_numbers:          1
  inodes_use_key_cache:         1
  gc_reserve_percent:           8
  gc_reserve_bytes:             0
  root_reserve_percent:         0
  wide_macs:                    0
  acl:                          1
  usrquota:                     0
  grpquota:                     0
  prjquota:                     0
  journal_flush_delay:          1000
  journal_flush_disabled:       0
  journal_reclaim_delay:        1000

members (size 344):
  Device:                       0
    UUID:                       8f6bbc2d-3b2b-4423-ad13-1a50108279e0
    Size:                       2.7T
    Bucket size:                512k
    First bucket:               0
    Buckets:                    5722222
    Last mount:                 Fri Jun  3 04:44:03 2022
    State:                      rw
    Label:                      rotational (0)
    Data allowed:               journal,btree,user
    Has data:                   journal,btree,user,cached
    Discard:                    0
    Freespace initialized:      1
  Device:                       1
    UUID:                       93b4098e-bba5-4577-8576-c2cd25086a6d
    Size:                       2.7T
    Bucket size:                512k
    First bucket:               0
    Buckets:                    5722222
    Last mount:                 Fri Jun  3 04:44:03 2022
    State:                      rw
    Label:                      rotational (0)
    Data allowed:               journal,btree,user
    Has data:                   journal,btree,user,cached
    Discard:                    0
    Freespace initialized:      1
  Device:                       2
    UUID:                       67701b0f-e143-4d48-8280-12dd5999b168
    Size:                       3.6T
    Bucket size:                512k
    First bucket:               0
    Buckets:                    7630895
    Last mount:                 Fri Jun  3 04:44:03 2022
    State:                      rw
    Label:                      rotational (0)
    Data allowed:               journal,btree,user
    Has data:                   journal,btree,user,cached
    Discard:                    0
    Freespace initialized:      1
  Device:                       3
    UUID:                       f7129d1d-25a9-4de4-9d16-e9fbd03d0629
    Size:                       3.6T
    Bucket size:                512k
    First bucket:               0
    Buckets:                    7630895
    Last mount:                 Fri Jun  3 04:44:03 2022
    State:                      rw
    Label:                      rotational (0)
    Data allowed:               journal,btree,user
    Has data:                   journal,btree,user,cached
    Discard:                    0
    Freespace initialized:      1
  Device:                       4
    UUID:                       1f0dfbdb-b149-4cc9-9b3e-f3624bee4b43
    Size:                       238G
    Bucket size:                256k
    First bucket:               0
    Buckets:                    976793
    Last mount:                 Fri Jun  3 04:44:03 2022
    State:                      rw
    Label:                      ssd (1)
    Data allowed:               journal,btree,user
    Has data:                   journal,btree,user,cached
    Discard:                    1
    Freespace initialized:      1
  Device:                       5
    UUID:                       0ceeb0bd-0b50-4b96-ab96-305c8380b3e9
    Size:                       238G
    Bucket size:                512k
    First bucket:               0
    Buckets:                    488396
    Last mount:                 Fri Jun  3 04:44:03 2022
    State:                      rw
    Label:                      ssd (1)
    Data allowed:               journal,btree,user
    Has data:                   journal,btree,user,cached
    Discard:                    1
    Freespace initialized:      1
</pre>

[Log of gdb session](https://github.com/koverstreet/bcachefs/files/8836988/gdb.log)

End of log:
```
bcachefs: libbcachefs/btree_io.c:1845: __bch2_btree_node_write: Assertion `!(memcmp(&b->data->format, &b->format, sizeof(b->format)))' failed.

Thread 1 "bcachefs" received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737349667712) at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737349667712) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737349667712) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737349667712, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff7c03476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff7be97f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff7be971b in __assert_fail_base (fmt=0x7ffff7d9e150 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x555555650200 "!(memcmp(&b->data->format, &b->format, sizeof(b->format)))", 
    file=0x55555564f70b "libbcachefs/btree_io.c", line=1845, function=<optimized out>) at ./assert/assert.c:92
#6  0x00007ffff7bfae96 in __GI___assert_fail (assertion=assertion@entry=0x555555650200 "!(memcmp(&b->data->format, &b->format, sizeof(b->format)))", file=file@entry=0x55555564f70b "libbcachefs/btree_io.c", line=line@entry=1845, 
    function=function@entry=0x555555650550 <__PRETTY_FUNCTION__.4> "__bch2_btree_node_write") at ./assert/assert.c:101
#7  0x00005555555f8425 in __bch2_btree_node_write (c=c@entry=0x7ffff7a9b000, b=b@entry=0x55565936b600, flags=flags@entry=0) at libbcachefs/btree_io.c:1845
#8  0x00005555555f96e9 in bch2_btree_node_write (c=0x7ffff7a9b000, b=0x55565936b600, lock_type_held=SIX_LOCK_write, flags=0) at libbcachefs/btree_io.c:2105
#9  0x00005555555f9a46 in bch2_btree_init_next (trans=trans@entry=0x7fffffffdc40, b=b@entry=0x55565936b600) at libbcachefs/btree_io.c:480
#10 0x00005555555da626 in bch2_btree_node_prep_for_write (b=0x55565936b600, path=<optimized out>, trans=0x7fffffffdc40) at libbcachefs/btree_update_leaf.c:75
#11 trans_lock_write (trans=0x7fffffffdc40) at libbcachefs/btree_update_leaf.c:826
#12 do_bch2_trans_commit (trace_ip=<optimized out>, stopped_at=<synthetic pointer>, trans=0x7fffffffdc40) at libbcachefs/btree_update_leaf.c:918
#13 __bch2_trans_commit (trans=trans@entry=0x7fffffffdc40) at libbcachefs/btree_update_leaf.c:1154
#14 0x000055555558e33a in bch2_trans_commit (flags=<optimized out>, journal_seq=0x0, disk_res=0x0, trans=0x7fffffffdc40) at libbcachefs/btree_update.h:106
#15 bch2_journal_replay (c=c@entry=0x7ffff7a9b000) at libbcachefs/recovery.c:647
#16 0x0000555555590517 in bch2_fs_recovery (c=0x7ffff7a9b000) at libbcachefs/recovery.c:1263
#17 0x000055555557de3d in bch2_fs_start (c=c@entry=0x7ffff7a9b000) at libbcachefs/super.c:891
#18 0x00005555555806ed in bch2_fs_open (devices=devices@entry=0x7fffffffe590, nr_devices=nr_devices@entry=6, opts=...) at libbcachefs/super.c:1898
#19 0x0000555555619a15 in cmd_fsck (argc=6, argc@entry=9, argv=0x7fffffffe590, argv@entry=0x7fffffffe578) at cmd_fsck.c:90
#20 0x000055555555ee1e in main (argc=<optimized out>, argv=0x7fffffffe578) at bcachefs.c:198
(gdb) up
#1  __pthread_kill_internal (signo=6, threadid=140737349667712) at ./nptl/pthread_kill.c:78
78	in ./nptl/pthread_kill.c
(gdb) 
#2  __GI___pthread_kill (threadid=140737349667712, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
89	in ./nptl/pthread_kill.c
(gdb) 
#3  0x00007ffff7c03476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
26	../sysdeps/posix/raise.c: No such file or directory.
(gdb) 
#4  0x00007ffff7be97f3 in __GI_abort () at ./stdlib/abort.c:79
79	./stdlib/abort.c: No such file or directory.
(gdb) 
#5  0x00007ffff7be971b in __assert_fail_base (fmt=0x7ffff7d9e150 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x555555650200 "!(memcmp(&b->data->format, &b->format, sizeof(b->format)))", 
    file=0x55555564f70b "libbcachefs/btree_io.c", line=1845, function=<optimized out>) at ./assert/assert.c:92
92	./assert/assert.c: No such file or directory.
(gdb) 
#6  0x00007ffff7bfae96 in __GI___assert_fail (assertion=assertion@entry=0x555555650200 "!(memcmp(&b->data->format, &b->format, sizeof(b->format)))", file=file@entry=0x55555564f70b "libbcachefs/btree_io.c", line=line@entry=1845, 
    function=function@entry=0x555555650550 <__PRETTY_FUNCTION__.4> "__bch2_btree_node_write") at ./assert/assert.c:101
101	in ./assert/assert.c
(gdb) 
#7  0x00005555555f8425 in __bch2_btree_node_write (c=c@entry=0x7ffff7a9b000, b=b@entry=0x55565936b600, flags=flags@entry=0) at libbcachefs/btree_io.c:1845
1845		BUG_ON(memcmp(&b->data->format, &b->format, sizeof(b->format)));
(gdb) print b->data->format
$1 = {key_u64s = 233 '\351', nr_fields = 193 '\301', bits_per_field = "\373-Y1\255\035", field_offset = {8967320364902175699, 3341292044210367861, 14373201367735718771, 16120446493376398537, 13125983515732037482, 15600305306382283820}}
(gdb) print b->format
$2 = {key_u64s = 2 '\002', nr_fields = 6 '\006', bits_per_field = "\000\020 \000\000", field_offset = {0, 1074155139, 0, 0, 0, 0}}
```

**Comment by RAOF** [comment]
With additional asserts:
```
diff --git a/libbcachefs/btree_iter.c b/libbcachefs/btree_iter.c
index 0515625..f84031c 100644
--- a/libbcachefs/btree_iter.c
+++ b/libbcachefs/btree_iter.c
@@ -140,6 +140,8 @@ static inline bool btree_path_pos_in_node(struct btree_path *path,
 void bch2_btree_node_unlock_write(struct btree_trans *trans,
                        struct btree_path *path, struct btree *b)
 {
+       if (b->data)
+               BUG_ON(memcmp(&b->data->format, &b->format, sizeof(b->format)));
        bch2_btree_node_unlock_write_inlined(trans, path, b);
 }
 
@@ -148,6 +150,8 @@ void __bch2_btree_node_lock_write(struct btree_trans *trans, struct btree *b)
        struct btree_path *linked;
        unsigned readers = 0;
 
+       if (b->data)
+               BUG_ON(memcmp(&b->data->format, &b->format, sizeof(b->format)));
        trans_for_each_path(trans, linked)
                if (linked->l[b->c.level].b == b &&
                    btree_node_read_locked(linked, b->c.level))
@@ -369,6 +373,8 @@ bool __bch2_btree_node_lock(struct btree_trans *trans,
                }
        }
 
+       if (b->data)
+               BUG_ON(memcmp(&b->data->format, &b->format, sizeof(b->format)));
        return btree_node_lock_type(trans, path, b, pos, level,
                                    type, should_sleep_fn, p);
 deadlock:
```

We instead get:
```
ptr: u64s 12 type btree_ptr_v2 0:1074156450:U32_MAX len 0 ver 0: seq 70e37f33462757d2 written 408 min_key 0:1074155139:0 ptr: 5:357065:0 gen 21 ptr: 4:804374:0 gen 20
header: btree subvolumes level 7
min 5374773745560097295:5129301685385335711:3931369734
max 9659435150497242846:8633935188037142342:129379454
btree node header doesn't match ptr
btree inodes level 0
ptr: u64s 12 type btree_ptr_v2 0:1074156450:U32_MAX len 0 ver 0: seq 70e37f33462757d2 written 408 min_key 0:1074155139:0 ptr: 5:357065:0 gen 21 ptr: 4:804374:0 gen 20
header: btree subvolumes level 7
min 5374773745560097295:5129301685385335711:3931369734
max 9659435150497242846:8633935188037142342:129379454
btree node header doesn't match ptr
btree inodes level 0
ptr: u64s 12 type btree_ptr_v2 0:1074156450:U32_MAX len 0 ver 0: seq 70e37f33462757d2 written 408 min_key 0:1074155139:0 ptr: 5:357065:0 gen 21 ptr: 4:804374:0 gen 20
header: btree subvolumes level 7
min 5374773745560097295:5129301685385335711:3931369734
max 9659435150497242846:8633935188037142342:129379454
bcachefs: libbcachefs/btree_io.c:1845: __bch2_btree_node_write: Assertion `!(memcmp(&b->data->format, &b->format, sizeof(b->format)))' failed.
Aborted
```

**Comment by RAOF** [comment]
This was fixed by 8736a249ab4f5925dc8789cb09dc9884578f3527
